docs/architecture/adr-101-data-companion-pull-api.md

```diff
@@ -0,0 +1,397 @@
+
+message GetBlockResultsRetainHeightResponse {
+    // The retain height set by the pruning service (e.g. by the data
+    // companion) specifically for block results.
+    uint64 pruning_service_retain_height = 1;
+}
+```
+
+With this API design, it is technically possible for an integrator to attach
+multiple data companions to the node, but only one of their retain heights will
+be considered by the node.
+
+### Configuration
+
+The following configuration file update is proposed to support the data
+companion API.
+
+```toml
+[storage]
+
+#
+# Storage pruning configuration relating only to the data companion.
+#
+[storage.pruning.data_companion]
+
+# Whether automatic pruning respects values set by the data companion. Disabled
+# by default. All other parameters in this section are ignored when this is
+# disabled.
+#
+# If disabled, only the application retain height will influence block pruning
+# (but not block results pruning). Only enabling this at a later stage will
+# potentially mean that blocks below the application-set retain height at the
+# time will not be available to the data companion.
+enabled = false
+
+# The initial value for the data companion block retain height if the data
+# companion has not yet explicitly set one. If the data companion has already
+# set a block retain height, this is ignored.
+initial_block_retain_height = 0
+
+# The initial value for the data companion block results retain height if the
+# data companion has not yet explicitly set one. If the data companion has
+# already set a block results retain height, this is ignored.
+initial_block_results_retain_height = 0
+
+#
+# This is the envisaged configuration section for the gRPC API that will be
+# introduced as part of https://github.com/cometbft/cometbft/issues/81
+# (Still a WIP)
+#
+[grpc]
+
+# The host/port on which to expose non-privileged gRPC endpoints.
+laddr = "tcp://localhost:26654"
+
+#
+# Configuration for privileged gRPC endpoints, which should **never** be exposed
+# to the public internet.
+#
+[grpc.privileged]
+# The host/port on which to expose privileged gRPC endpoints.
+laddr = "tcp://localhost:26655"
+
+#
+# Configuration specifically for the gRPC pruning service, which is considered a
+# privileged service.
+#
+[grpc.privileged.pruning_service]
+
+# Only controls whether the pruning service is accessible via the gRPC API - not
+# whether a previously set pruning service retain height is honoured by the
+# node. That is controlled by settings in the [storage.pruning] section.
+#
+# Disabled by default.
+enabled = false
+```
+
+### Metrics
+
+The following metrics are proposed to be added to monitor the health of the
+interaction between a node and its data companion:
+
+- `pruning_service_block_retain_height` - The current block retain height as
+  requested by the pruning service.
+- `pruning_service_block_results_retain_height` - The current block results
+  retain height as requested by the pruning service.
+
```

Comment:
I propose we extend this list with the following:
 - `application_block_retain_height`
 - `block_store_base_height`
 - `abci_results_base_height`
 
 The first one I am suggesting because this allows an operator to see whether there is currently a lower retain height set and why the blocks the data companion wants pruned are not pruned. 
 The other two are there to indicate how far has pruning gotten. I also had a conversation with Milan today where he mentioned that sometimes they are not aware a block is not there and they try to get a block etc.

Reply:
I'll update the ADR accordingly!

Reply:
Updated